src/lightbox/downloadImage.js

```diff
@@ -34,28 +34,45 @@ const androidEnsureStoragePermission = async (): Promise<void> => {
   // result === GRANTED
 };
 
-export default async (src: string, auth: Auth): Promise<mixed> => {
-  const absoluteUrl = getFullUrl(src, auth.realm);
+export default async (
+  src: string,
+  auth: Auth,
+): Promise<{|
+  url: string,
+  openedInBrowser: boolean,
+|}> => {
+  const tempUrl = await tryGetTemporaryFileUrl(src, auth);
+
+  if (tempUrl === null) {
+    const fullUrl = getFullUrl(src, auth.realm);
+    openLink(fullUrl);
+    return {
+      url: fullUrl,
+      openedInBrowser: true,
+    };
+  }
 
   if (Platform.OS === 'ios') {
-    const delimiter = absoluteUrl.includes('?') ? '&' : '?';
-    const urlWithApiKey = `${absoluteUrl}${delimiter}api_key=${auth.apiKey}`;
-    return CameraRoll.saveToCameraRoll(urlWithApiKey);
+    return {
+      url: await CameraRoll.saveToCameraRoll(tempUrl),
+      openedInBrowser: false,
+    };
   }
 
   // Platform.OS === 'android'
```

Comment:
Hmm, all this logic is kind of twisty to follow.

One thing that I notice is that both of this function's callers want to inspect the `openedInBrowser` value... and that in turn corresponds precisely to a fork in the function's control flow, where it does one thing in one case and a disjoint set of things in the other case.

As a result I suspect both this function, and its callers, could be made easier to understand by pushing that condition out into the callers. Let's try that.

Reply:
I understand that you want the `downloadImage` export to only download the file, and throw if it fails. The callers should handle the exception themselves, and open the file in the browser if required. Is this correct?

Reply:
Or, we can also pass a valid temporary URL to this function in the first place, and it should never throw. The caller will handle the generation of the temp URL and the failure to do that. This sounds better.

Reply:
I went ahead with the 2nd approach.

Reply:
Yep, that second approach is what I had in mind :slightly_smiling_face: -- looks good!